crates/matrix-sdk-ui/src/timeline/inner.rs

```diff
@@ -958,3 +1215,88 @@ async fn fetch_replied_to_event(
     };
     Ok(res)
 }
+
+fn update_timeline_reaction(
+    mut state: MutexGuard<'_, TimelineInnerState>,
+    own_user_id: &UserId,
+    annotation: &Annotation,
+    result: &ReactionToggleResult,
+) -> Result<(), super::Error> {
+    if matches!(result, ReactionToggleResult::RedactSuccess) {
+        // We did a successful redaction, so no need to update the item
+        // because the reaction is already gone.
+        return Ok(());
+    }
+
+    let (remote_echo_to_add, local_echo_to_remove) = match result {
+        ReactionToggleResult::AddSuccess { event_id, txn_id } => (Some(event_id), Some(txn_id)),
+        ReactionToggleResult::AddFailure { txn_id } => (None, Some(txn_id)),
+        ReactionToggleResult::RedactSuccess => (None, None),
+        ReactionToggleResult::RedactFailure { event_id } => (Some(event_id), None),
+    };
+
+    let related = rfind_event_item(&state.items, |it| {
+        it.event_id().is_some_and(|it| it == annotation.event_id)
+    });
+
+    let Some((idx, related)) = related else {
+        // Event isn't found at all.
+        warn!("Timeline item not found, can't update reaction ID");
+        return Err(super::Error::FailedToToggleReaction);
+    };
+    let Some(remote_related) = related.as_remote() else {
+        error!("inconsistent state: reaction received on a non-remote event item");
+        return Err(super::Error::FailedToToggleReaction);
+    };
+
+    let new_reactions = {
+        let mut reactions = remote_related.reactions.clone();
+        let reaction_group = reactions.entry(annotation.key.clone()).or_default();
```

Comment:
Can we again keep the entry in a local variable here, so we can use `entry.remove()` below (line 1271) instead of using the map's `remove` which has to look up the same key again?

Reply:
I see what you mean. Let's keep it as-is, I can refactor later if I find a way.